### ISSUE TITLE
error: no matching function for call to ‘ConcreteVectorArithmetic()’

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#2073]
I've seen the following on a number of PPA builds:

* [Ubuntu 18.04, arm64](https://launchpadlibrarian.net/538549516/buildlog_ubuntu-bionic-arm64.macaulay2_1.17.2.1+git202105131229-0ppa202105121825~ubuntu18.04.1_BUILDING.txt.gz)
* [Ubuntu 18.04, i386](https://launchpadlibrarian.net/538549092/buildlog_ubuntu-bionic-i386.macaulay2_1.17.2.1+git202105131229-0ppa202105121825~ubuntu18.04.1_BUILDING.txt.gz)
* [Ubuntu 18.04, ppc64el](https://launchpadlibrarian.net/538549331/buildlog_ubuntu-bionic-ppc64el.macaulay2_1.17.2.1+git202105131229-0ppa202105121825~ubuntu18.04.1_BUILDING.txt.gz)

```
g++ -pipe -Wall -Wno-shadow -Wcast-qual -Wno-sign-conversion -Wno-sign-compare -Wno-parentheses -Wno-sign-compare  -fno-stack-protector -Wuninitialized -I/usr/include  -Wno-error=pragmas -std=gnu++14 -g -fdebug-prefix-map=/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229=. -fstack-protector-strong -Wformat -Werror=format-security -g3 -O2  -Wno-unused-local-typedefs -Wno-unknown-pragmas -std=gnu++17 -Wno-mismatched-tags -I. -DNDEBUG -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/usr-host/include  -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/usr-host/include  -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/include -I/<<BUILDDIR>>/macaulay2-1.17.2.1+git202105131229/M2/usr-host/include  -Wdate-time -D_FORTIFY_SOURCE=2 -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular -I/usr/include/powerpc64le-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -I../d -I./../c -Wno-fatal-errors  -c -Wno-cast-qual NCAlgebras/NCF4.cpp -o NCAlgebras/NCF4.o
In file included from ./NCAlgebras/NCF4.hpp:18:0,
                 from NCAlgebras/NCF4.cpp:1:
./VectorArithmetic.hpp: In constructor ‘VectorArithmetic::VectorArithmetic(const Ring*)’:
./VectorArithmetic.hpp:471:85: error: class template argument deduction failed:
           (R, &dynamic_cast< const M2::ConcreteRing<M2::ARingZZpFlint>* >(R)->ring());
                                                                                     ^
./VectorArithmetic.hpp:471:85: error: no matching function for call to ‘ConcreteVectorArithmetic()’
./VectorArithmetic.hpp:99:3: note: candidate: template<class RingType, class CoeffVectorType, class DenseCoeffVectorType> ConcreteVectorArithmetic(const Ring*, const RingType*)-> ConcreteVectorArithmetic<RingType, CoeffVectorType, DenseCoeffVectorType>
   ConcreteVectorArithmetic(const Ring* origR, const RingType* R) : mOriginalRing(origR), mRing(R) {}
   ^~~~~~~~~~~~~~~~~~~~~~~~
./VectorArithmetic.hpp:99:3: note:   template argument deduction/substitution failed:
./VectorArithmetic.hpp:471:85: note:   candidate expects 2 arguments, 0 provided
           (R, &dynamic_cast< const M2::ConcreteRing<M2::ARingZZpFlint>* >(R)->ring());
                                                                                     ^
./VectorArithmetic.hpp:66:3: note: candidate: template<class RingType, class CoeffVectorType, class DenseCoeffVectorType> ConcreteVectorArithmetic()-> ConcreteVectorArithmetic<RingType, CoeffVectorType, DenseCoeffVectorType>
   ConcreteVectorArithmetic() : mOriginalRing(nullptr), mRing(nullptr) {}
   ^~~~~~~~~~~~~~~~~~~~~~~~
./VectorArithmetic.hpp:66:3: note:   template argument deduction/substitution failed:
./VectorArithmetic.hpp:471:85: note:   couldn't deduce template parameter ‘RingType’
           (R, &dynamic_cast< const M2::ConcreteRing<M2::ARingZZpFlint>* >(R)->ring());
                                                                                     ^
./VectorArithmetic.hpp:475:87: error: class template argument deduction failed:
           (R, &dynamic_cast< const M2::ConcreteRing<M2::ARingGFFlintBig>* >(R)->ring());
                                                                                       ^
./VectorArithmetic.hpp:475:87: error: no matching function for call to ‘ConcreteVectorArithmetic()’
./VectorArithmetic.hpp:99:3: note: candidate: template<class RingType, class CoeffVectorType, class DenseCoeffVectorType> ConcreteVectorArithmetic(const Ring*, const RingType*)-> ConcreteVectorArithmetic<RingType, CoeffVectorType, DenseCoeffVectorType>
   ConcreteVectorArithmetic(const Ring* origR, const RingType* R) : mOriginalRing(origR), mRing(R) {}
   ^~~~~~~~~~~~~~~~~~~~~~~~

etc.
```

**Comment by mikestillman** [comment]
@d-torrance, do you know what compilers are being used?

**Comment by d-torrance** [comment]
> @d-torrance, do you know what compilers are being used?

```
checking for gcc version... 7
```

All the Ubuntu 18.04 builds have failed now, so I think it's definitely a GCC 7 thing.

**Comment by d-torrance** [comment]
GCC 8 is available in 18.04 -- I'll see if using that fixes it.

**Comment by mikestillman** [comment]
We are using a c++-17 feature, which might not be supported by that compiler, I suppose.  @moorewf Any suggestions?

**Comment by mikestillman** [comment]
> GCC 8 is available in 18.04 -- I'll see if using that fixes it.

That sounds good!  We could remove use of that feature, but it would take a bit of work, and it would be nice to keep that feature...

**Comment by d-torrance** [comment]
Nope, didn't work with GCC 8 either.  :(

```
g++-8 -pipe -Wall -Wno-shadow -Wcast-qual -Wno-sign-conversion -Wno-sign-compare -Wno-parentheses -Wno-sign-compare  -fno-stack-protector -Wuninitialized -I/usr/include  -Wno-error=pragmas -std=gnu++14 -g3 -O2  -Wno-unused-local-typedefs -Wno-unknown-pragmas -std=gnu++17 -Wno-mismatched-tags -I../../../../Macaulay2/e -DNDEBUG -I/root/M2/M2/BUILD/doug/../../include -I/root/M2/M2/BUILD/doug/include -I/root/M2/M2/BUILD/doug/usr-host/include  -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular   -I/usr/src/gtest   -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include -DBOOST_STACKTRACE_LINK -I../d -I../../../../Macaulay2/e/../c -Wno-fatal-errors  -c -Wno-cast-qual ../../../../Macaulay2/e/NCAlgebras/NCF4.cpp -o NCAlgebras/NCF4.o
In file included from ../../../../Macaulay2/e/NCAlgebras/NCF4.hpp:18,
                 from ../../../../Macaulay2/e/NCAlgebras/NCF4.cpp:1:
../../../../Macaulay2/e/VectorArithmetic.hpp: In constructor 'VectorArithmetic::VectorArithmetic(const Ring*)':
../../../../Macaulay2/e/VectorArithmetic.hpp:471:85: error: class template argument deduction failed:
    (R, &dynamic_cast< const M2::ConcreteRing<M2::ARingZZpFlint>* >(R)->ring());
                                                                              ^

../../../../Macaulay2/e/VectorArithmetic.hpp:471:85: error: no matching function for call to 'ConcreteVectorArithmetic()'
../../../../Macaulay2/e/VectorArithmetic.hpp:99:3: note: candidate: 'template<class RingType, class CoeffVectorType, class DenseCoeffVectorType> ConcreteVectorArithmetic(const Ring*, const RingType*)-> ConcreteVectorArithmetic<RingType, CoeffVectorType, DenseCoeffVectorType>'
   ConcreteVectorArithmetic(const Ring* origR, const RingType* R) : mOriginalRing(origR), mRing(R) {}
   ^~~~~~~~~~~~~~~~~~~~~~~~
../../../../Macaulay2/e/VectorArithmetic.hpp:99:3: note:   template argument deduction/substitution failed:
../../../../Macaulay2/e/VectorArithmetic.hpp:471:85: note:   candidate expects 2 arguments, 0 provided
    (R, &dynamic_cast< const M2::ConcreteRing<M2::ARingZZpFlint>* >(R)->ring());
                                                                              ^
```
Perhaps there's some compiler flags or something that will help.  I'll play around with it some more.

If push comes to shove, I suppose I could just stop building Ubuntu 18.04 packages.  It is two LTS releases ago, after all.

**Comment by moorewf** [comment]
Is [this](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85883) relevant?  Not sure if GCC 9 is available in 18.04 though.

**Comment by d-torrance** [comment]
> Not sure if GCC 9 is available in 18.04 though.

It isn't by default, but I did find [this PPA](https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test) that has it (and GCC 11, in fact!).  So I'll be able to use that and keep building 18.04 packages.

Maybe we should add a little test to `configure` to see if this bug is an issue?

**Comment by d-torrance** [comment]
Actually, the simple workaround suggested in [this comment on that bug report](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85883#c2) (use list initialization, i.e., just switch parentheses to curly braces) seems to do the trick!

I'll open a PR here in a moment.
